### PR TITLE
Parallel verification of signatures in a block

### PR DESCRIPTION
**Pull request by tock-ibm**
Validate the signatures on transactions in a block in parallel rather then serially, for performance.

First validate the signatures on each transaction in parallel, remembering the
set of users with valid signatures per TX. Then validate the fields, access control, operations, and MVCC serially.

`t.Parallel()` was removed from `internal/bcdb` unit tests because it creates an out-of-memory failure due to too many tests with `leveldb` running together.

Signed-off-by: Yoav Tock <tock@il.ibm.com>

### CODE REVIEW

File: internal/bcdb/ledger_query_procesor_test.go [hunk 1]
```diff
@@ -370,7 +370,6 @@ func createDataUpdatesFromBlock(block *types.Block) map[string]*worldstate.DBUpd
 }
 
 func TestGetBlock(t *testing.T) {
-	t.Parallel()
```
Comment (C0rWin) [inline internal/bcdb/ledger_query_procesor_test.go:373]: Why did you remove this?

File: internal/blockprocessor/data_tx_validator.go [hunk 2]
```diff
@@ -99,7 +75,33 @@ func (v *dataTxValidator) validate(txEnv *types.DataTxEnvelope, pendingOps *pend
...
+		if err != nil {
+			return nil, nil, err
+		}
+		if valRes.Flag != types.Flag_VALID {
+			for _, mustSignUserID := range txEnv.Payload.MustSignUserIds {
+				if userID == mustSignUserID {
+					return nil,
+						&types.ValidationInfo{
+							Flag:            types.Flag_INVALID_UNAUTHORISED,
+							ReasonIfInvalid: "signature of the must sign user [" + userID + "] is not valid (maybe the certifcate got changed)",
```
Comment (C0rWin) [inline internal/blockprocessor/data_tx_validator.go:94]:
```suggestion
							ReasonIfInvalid: "signature of the must sign user [" + userID + "] is not valid (maybe the certificate got changed)",
```

File: internal/blockprocessor/validator.go [hunk 3]
```diff
@@ -182,6 +190,40 @@ func (v *validator) validateBlock(block *types.Block) ([]*types.ValidationInfo,
 	}
 }
 
+func (v *validator) parallelSigValidation(dataTxEnvs []*types.DataTxEnvelope) ([]*types.ValidationInfo, [][]string, error) {
+	valInfoArray := make([]*types.ValidationInfo, len(dataTxEnvs))
+	usersWithValidSigPerTX := make([][]string, len(dataTxEnvs))
+	errorArray := make([]error, len(dataTxEnvs))
```
Comment (cendhu) [inline internal/blockprocessor/validator.go:196]: Would be good to avoid the type name in the variable.

```suggestion
	txsValInfo := make([]*types.ValidationInfo, len(dataTxEnvs))
	usersWithValidSigPerTX := make([][]string, len(dataTxEnvs))
	txsValError := make([]error, len(dataTxEnvs))
```

File: internal/blockprocessor/validator.go [hunk 4]
```diff
@@ -182,6 +190,40 @@ func (v *validator) validateBlock(block *types.Block) ([]*types.ValidationInfo,
...
+			} else {
+				*usersWithValidSig = usersWithValidSignTx
+				*valInfo = vInfo
+				if vInfo.Flag != types.Flag_VALID {
+					v.logger.Debugf("data transaction [%v] is invalid due to [%s]", txEnv.Payload, vInfo.ReasonIfInvalid)
+				}
+			}
+			wg.Done()
+		}(txEnv, &usersWithValidSigPerTX[txNum], &valInfoArray[txNum], &errorArray[txNum])
+	}
```
Comment (cendhu) [inline internal/blockprocessor/validator.go:215]: The following can be used to improve code readability (it does not create any race condition too). 

```suggestion
		go func(txEnv *types.DataTxEnvelope, txNum int) {
			usersWithValidSignTx, vInfo, vErr := v.dataTxValidator.validateSignatures(txEnv)
			if vErr != nil {
				errorArray[txNum] = err
			} else {
				usersWithValidSigPerTx[txNum] = usersWithValidSignTx
				valInfoArray[txNum] = vInfo
				if vInfo.Flag != types.Flag_VALID {
					v.logger.Debugf("data transaction [%v] is invalid due to [%s]", txEnv.Payload, vInfo.ReasonIfInvalid)
				}
			}
			wg.Done()
		}(txEnv, txNum)
	}
```

Comment (cendhu) [reply internal/blockprocessor/validator.go:215]: Nit: we can avoid the extra indentation. 

```
		go func(txEnv *types.DataTxEnvelope, txNum int) {
			defer wg.Done()
			
                        usersWithValidSignTx, vInfo, vErr := v.dataTxValidator.validateSignatures(txEnv)
			if vErr != nil {
				errorArray[txNum] = vErr
				return
			}

			usersWithValidSigPerTX[txNum] = usersWithValidSignTx
			valInfoArray[txNum] = vInfo
			if vInfo.Flag != types.Flag_VALID {
				v.logger.Debugf("data transaction [%v] is invalid due to [%s]", txEnv.Payload, vInfo.ReasonIfInvalid)
			}
		}(txEnv, txNum)

```

File: internal/bcdb/ledger_query_procesor_test.go (hunk 1 above)
Comment (tock-ibm) [reply internal/bcdb/ledger_query_procesor_test.go:373]: It causes the `interenal/bcdb` unit tests to throw an OOM... too many tests together. Even if I run just the BCDB tests, not the entire suite...

File: internal/blockprocessor/data_tx_validator.go (hunk 2 above)
Comment (tock-ibm) [reply internal/blockprocessor/data_tx_validator.go:94]: Done

File: internal/blockprocessor/validator.go (hunk 3 above)
Comment (tock-ibm) [reply internal/blockprocessor/validator.go:196]: After `usersWithValidSigPerTX` I changed it to `valInfoPerTx` & `errorPerTx`; now it's all nice and tidy.

File: internal/blockprocessor/validator.go (hunk 4 above)
Comment (tock-ibm) [reply internal/blockprocessor/validator.go]: Done

File: internal/blockprocessor/validator.go [hunk 5]
```diff
@@ -104,16 +105,23 @@ func (v *validator) validateBlock(block *types.Block) ([]*types.ValidationInfo,
 	switch block.Payload.(type) {
 	case *types.Block_DataTxEnvelopes:
 		dataTxEnvs := block.GetDataTxEnvelopes().Envelopes
-		valInfo := make([]*types.ValidationInfo, len(dataTxEnvs))
-		pendingOps := newPendingOperations()
+		valInfoArray, usersWithValidSigPerTX, err := v.parallelSigValidation(dataTxEnvs)
```
Comment (cendhu) [inline internal/blockprocessor/validator.go:108]: could change `valInfoArray` to `valInfoPerTx`.

**Review by cendhu — APPROVED** [review]
